src/test/app/Offer_test.cpp

```diff
@@ -795,22 +850,831 @@ class Offer_test : public beast::unit_test::suite
             owners ("eve", 1));
     }
 
+    void
+    testcaseSelfCross()
+    {
+        testcase ("Self-crossing");
+
+        using namespace jtx;
+
+        auto const gw = Account {"gateway"};
+        auto const USD = gw["USD"];
+        auto const BTC = gw["BTC"];
+
+        Env env {*this};
+        env.fund (XRP (10000), gw);
+        env (offer (gw, BTC (500), USD (100)));
```

Comment:
Hmm. While it's neat to test the gateway trading its own assets, I'd also like to see a test where the gateway funds an account and it does the trading. Why does this matter? An issuer always has an unlimited amount of its own funds.

Also, I'd recommend performing two variants of this test: one is what you have BTC-USD with BTC and USD offers. The other is BTC-USD, but the offers placed are BTC-XRP and XRP-USD, so that you autobridge. If you want, you can combine them, by having half the offer be direct BTC-USD, and the other hand bridge BTC-XRP and XRP-USD (you just need to be careful how you place the orders: BTC-XRP, BTC-USD and XRP-USD).